### PR TITLE
feat: repo unit tests

### LINKED ISSUE #66: Repo service route testing
We should add some unit tests for routing

### PR DESCRIPTION
**Pull request by rjzak**
Closes #66 
Signed-off-by: Richard Zak <richard@profian.com>

### CODE REVIEW

File: crates/repo/src/lib.rs [hunk 1]
```diff
@@ -228,6 +165,10 @@ pub fn app() -> Router {
 #[cfg(test)]
 mod tests {
     use super::*;
+    use http::header;
+    use hyper::client;
+    use namespace::Namespace;
+    use std::net::{SocketAddr, TcpListener};
 
     #[test]
     fn namespace_from_str() {
```
Comment (rvolosatovs) [inline crates/repo/src/lib.rs:174]: This test should be moved to `namespace.rs` and renamed to `from_str`

**Review by rvolosatovs — CHANGES_REQUESTED** [review]
Can we have 2 commits here:
1. Extract `namespace` into a separate module
2. Add routing tests
?

File: crates/repo/src/namespace.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
+// SPDX-FileCopyrightText: 2022 Profian Inc. <opensource@profian.com>
+// SPDX-License-Identifier: Apache-2.0
+
+#![warn(unused_lifetimes, unused_qualifications, clippy::all)]
+#![forbid(unsafe_code)]
```
Comment (rvolosatovs) [inline crates/repo/src/namespace.rs:5]:
```suggestion
```

File: crates/repo/src/lib.rs [hunk 3]
```diff
@@ -279,4 +220,40 @@ mod tests {
...
+
+    #[tokio::test]
+    async fn routes() {
+        let listener = TcpListener::bind("127.0.0.1:8000".parse::<SocketAddr>().unwrap()).unwrap();
+        let addr = listener.local_addr().unwrap();
+
+        tokio::spawn(async move {
+            axum::Server::from_tcp(listener)
+                .unwrap()
+                .serve(app().into_make_service())
```
Comment (rvolosatovs) [inline crates/repo/src/lib.rs:232]: What if port `8000` is already used?

Let's do exactly what we already do in main with no custom logic
https://github.com/profianinc/drawbridge/blob/19f8f307bcd7b30f921aa3d96a2876bca0cb73c6/src/main.rs#L14

File: crates/repo/src/lib.rs [hunk 4]
```diff
@@ -279,4 +220,40 @@ mod tests {
...
+                    .header(header::CONTENT_LENGTH, "16")
+                    .uri(format!("http://{}/test/onetwothree", addr))
+                    .body(Body::empty())
+                    .unwrap(),
+            )
+            .await
+            .unwrap();
+
+        // TODO: Validate output for successful namespace creation
+        let body = hyper::body::to_bytes(response.into_body()).await.unwrap();
```
Comment (rvolosatovs) [inline crates/repo/src/lib.rs:257]: What I would expect to happen here, would be a `GET` on non-existing repo, asserting a 404, then `PUT` to that repo asserting 200, then `GET` on the same repo now succeeding and asserting 200.

**Comment by rvolosatovs** [comment]
Also:
- please add "Closes #<issue number>" to PR descriptions of PRs closing issues. (just reference issues if they are related, but PR does not close them)
- ensure there's a space between the <type> in the commit message and the description, i.e. `chore: something` and *not* `chore:something`

File: crates/repo/src/lib.rs [hunk 5]
```diff
@@ -210,3 +210,42 @@ pub fn app() -> Router {
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use namespace::Namespace;
+    use std::net::{IpAddr, Ipv4Addr, SocketAddr, TcpListener};
+
+    #[tokio::test]
+    async fn routes() {
+        tokio::spawn(async move {
```
Comment (rjzak) [inline crates/repo/src/lib.rs:222]: @rvolosatovs The client fails to connect and I think there is an issue here getting the connection opened.

File: crates/repo/src/lib.rs [hunk 6]
```diff
@@ -210,3 +210,29 @@ pub fn app() -> Router {
         }),
     )
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use namespace::Namespace;
```
Comment (rvolosatovs) [inline crates/repo/src/lib.rs:217]: unused

**Review by rvolosatovs — CHANGES_REQUESTED** [review]
See https://github.com/profianinc/drawbridge/pull/68#discussion_r854962527

**Comment by rjzak** [comment]
I don't understand what happened, I rebased based on the changed upstream main, and now the upstream commits are now added to this PR?

**Comment by rvolosatovs** [comment]
> I don't understand what happened, I rebased based on the changed upstream main, and now the upstream commits are now added to this PR?

Let's try to go through the process together after standup

**Review by rvolosatovs — COMMENTED** [review]
A few improvements/fixes:
- `Content-Type` should not be specified on `GET` (although we should probably add `Accept` here later)
- `GET` is default, so do not specify it
- Added a `request_builder()` helper
- Removed the amount of assignments, so it's a bit clearer what we are asserting
- Imported the header constants
- Added `Content-Length` on `PUT`
- Added a few more test cases, especially error cases

```diff
diff --git a/crates/repo/src/lib.rs b/crates/repo/src/lib.rs
index 371deaa..78df45e 100644
--- a/crates/repo/src/lib.rs
+++ b/crates/repo/src/lib.rs
@@ -234,39 +234,94 @@ pub fn app() -> Router {
 mod tests {
     use super::*;
 
+    use axum::http::header::{CONTENT_LENGTH, CONTENT_TYPE};
+    use axum::http::request;
+
     #[tokio::test]
     async fn routes() {
-        const REPO_URL: &str = "http://127.0.0.1:8080/test/onetwothree";
+        const URL: &str = "http://127.0.0.1:8080/test/onetwothree";
+        fn request_builder() -> request::Builder {
+            Request::builder().uri(URL)
+        }
+
         let mut router = app();
 
-        let error_request = Request::builder()
-            .method("GET")
-            .header(axum::http::header::CONTENT_TYPE, "application/json")
-            .uri(REPO_URL)
-            .body(Body::empty())
+        let res = router
+            .call(request_builder().body(Body::empty()).unwrap())
+            .await
+            .unwrap();
+        assert_eq!(res.status(), StatusCode::NOT_FOUND);
+
+        let res = router
+            .call(
+                request_builder()
+                    .method("PUT")
+                    .header(CONTENT_LENGTH, 3)
+                    .header(CONTENT_TYPE, "application/json")
+                    .body(Body::from("{}"))
+                    .unwrap(),
+            )
+            .await
             .unwrap();
+        assert_eq!(res.status(), StatusCode::BAD_REQUEST);
 
-        let error_response = router.call(error_request).await.unwrap();
-        assert_eq!(error_response.status().as_u16(), 404);
+        let res = router
+            .call(
+                request_builder()
+                    .method("PUT")
+                    .header(CONTENT_LENGTH, 2)
+                    .header(CONTENT_TYPE, "application/toml")
+                    .body(Body::from("{}"))
+                    .unwrap(),
+            )
+            .await
+            .unwrap();
+        assert_eq!(res.status(), StatusCode::UNSUPPORTED_MEDIA_TYPE);
 
-        let put_request = Request::builder()
-            .method("PUT")
-            .header(axum::http::header::CONTENT_TYPE, "application/json")
-            .uri(REPO_URL)
-            .body(Body::from("{}"))
+        let res = router
+            .call(
+                request_builder()
+                    .method("PUT")
+                    .header(CONTENT_LENGTH, 13)
+                    .header(CONTENT_TYPE, "application/json")
+                    .body(Body::from(r#"{"one":"two"}"#))
+                    .unwrap(),
+            )
+            .await
             .unwrap();
+        assert_eq!(res.status(), StatusCode::UNPROCESSABLE_ENTITY);
 
-        let put_response = router.call(put_request).await.unwrap();
-        assert_eq!(put_response.status().as_u16(), 200);
+        let res = router
+            .call(
+                request_builder()
+                    .method("PUT")
+                    .header(CONTENT_LENGTH, 2)
+                    .header(CONTENT_TYPE, "application/json")
+                    .body(Body::from("{}"))
+                    .unwrap(),
+            )
+            .await
+            .unwrap();
+        assert_eq!(res.status(), StatusCode::OK);
 
-        let successful_request = Request::builder()
-            .method("GET")
-            .header(axum::http::header::CONTENT_TYPE, "application/json")
-            .uri(REPO_URL)
-            .body(Body::empty())
+        // TODO: This should not result in conflict, since payload is the same.
+        let res = router
+            .call(
+                request_builder()
+                    .method("PUT")
+                    .header(CONTENT_LENGTH, 2)
+                    .header(CONTENT_TYPE, "application/json")
+                    .body(Body::from("{}"))
+                    .unwrap(),
+            )
+            .await
             .unwrap();
+        assert_eq!(res.status(), StatusCode::CONFLICT);
 
-        let successful_response = router.call(successful_request).await.unwrap();
-        assert_eq!(successful_response.status().as_u16(), 200);
+        let res = router
+            .call(request_builder().body(Body::empty()).unwrap())
+            .await
+            .unwrap();
+        assert_eq!(res.status(), StatusCode::OK);
     }
 }
```

**Review by rvolosatovs — APPROVED** [review]

**Pull request merged**